File: Seminar/src/main/java/org/sopt/Seminar/global/common/exception/SoptException.java
package org.sopt.Seminar.global.common.exception;

import lombok.Getter;

@Getter
public class SoptException extends RuntimeException {
    int errorCode;
    ErrorType errorType;
    String detail;

    public SoptException(ErrorType errorType, String detail) {
        this.errorCode = errorType.getCode();
        this.errorType = errorType;
        this.detail = detail;
    }

    public abstract static class InvalidRequestException extends SoptException {
        public InvalidRequestException(ErrorType.BadRequest errorType, String detail) {
            super(errorType, detail);
        }
    }

    public abstract static class NotFoundException extends SoptException {
        public NotFoundException(ErrorType.NotFound errorType, String detail) {
            super(errorType, detail);
        }
    }

    public abstract static class ServerErrorException extends SoptException {
        public ServerErrorException(ErrorType.ServerError errorType, String detail) {
            super(errorType, detail);
        }
    }
}

File: kopring/week2/kotlin-semina/src/main/kotlin/org/sopt/kotlinsemina/global/common/exception/SoptException.kt
package org.sopt.kotlinsemina.global.common.exception

open class SoptException(
    errorType: ErrorType,
    detail: String
): RuntimeException() {
    val errorCode = errorType.errorCode
    val errorType = errorType
    val detail = detail

    abstract class InvalidRequestException(errorType: ErrorType.BadRequest, detail: String): SoptException(errorType, detail)

    abstract class NotFoundException(errorType: ErrorType.NotFound, detail: String): SoptException(errorType, detail)

    abstract class ServerErrorException(errorType: ErrorType.ServerError, detail: String): SoptException(errorType, detail)
}